immuni_analytics/tasks/authorize_analytics_token.py

```diff
@@ -0,0 +1,124 @@
+#   Copyright (C) 2020 Presidenza del Consiglio dei Ministri.
+#   Please refer to the AUTHORS file for more information.
+#   This program is free software: you can redistribute it and/or modify
+#   it under the terms of the GNU Affero General Public License as
+#   published by the Free Software Foundation, either version 3 of the
+#   License, or (at your option) any later version.
+#   This program is distributed in the hope that it will be useful,
+#   but WITHOUT ANY WARRANTY; without even the implied warranty of
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+#   GNU Affero General Public License for more details.
+#   You should have received a copy of the GNU Affero General Public License
+#   along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+import asyncio
+import random
+
+from immuni_analytics.celery import celery_app
+from immuni_analytics.core import config
+from immuni_analytics.core.managers import managers
+from immuni_analytics.helpers.device_check import fetch_device_check_bits, set_device_check_bits
+from immuni_analytics.helpers.redis import (
+    get_authorized_tokens_redis_key_current_month,
+    get_authorized_tokens_redis_key_next_month,
+)
+from immuni_common.core.exceptions import ImmuniException
+
+
+class DiscardAnalyticsTokenException(ImmuniException):
+    """Raised when the device cannot authorize an analytics token."""
+
+
+@celery_app.task()
+def authorize_analytics_token(analytics_token: str, device_token: str) -> None:  # pragma: no cover
+    """
+     Celery doesn't support async functions, so we wrap it around asyncio.run.
+     """
+    asyncio.run(_authorize_analytics_token(analytics_token, device_token))
+
+
+async def _authorize_analytics_token(analytics_token: str, device_token: str) -> None:
+    try:
+        # TODO wait time between one step and the next one
+        await _first_step(device_token)
+        await asyncio.sleep(random.uniform(1, config.CHECK_TIME))
+        await _second_step(device_token)
+        await asyncio.sleep(random.uniform(1, config.READ_TIME))
+        await _third_step(device_token)
+    except DiscardAnalyticsTokenException:
```

Comment:
is there any other possible exception than `DiscardAnalyticsTokenException` ?

Reply:
Should not be but worth a doublecheck